### PR TITLE
format: add support for gherkin rule/example syntax

### PR DESCRIPTION
**Pull request by davidjgoss**
After the move to Gherkin 9 and cucumber-messages, we can execute specs that include Rule/Example syntax, but still need to update the formatter helpers and built-in formatters to handle the results.

### CODE REVIEW

File: features/fixtures/formatters/passed-rule.message.json.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        ],
+      },
+    },
+  },
+  {
+    pickle: {
+      id: '4',
+      uri: 'features/a.feature',
+      name: 'an example',
+      language: 'en',
```
Comment (davidjgoss) [inline features/fixtures/formatters/passed-rule.message.json.ts]: Should the pickle name here be some kind of concatenation of the rule name and example name, or is the example name sufficient?

File: features/fixtures/formatters/passed-rule.json.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+module.exports = [
+  {
+    description: '',
+    elements: [
```
Comment (davidjgoss) [inline features/fixtures/formatters/passed-rule.json.ts]: Given the JSON formatter’s status as soon-to-be-deprecated, and other tools’ reliance on it, would seem best to not add any additional structure for Rule/Example, and instead just fit within the roaring structure for scenarios - do we think that’s right?

**Comment by davidjgoss** [comment]
The happy path for JSON formatter is working, moving on to failure cases and other built-in formatters, a couple of bits I am unsure of (above) which can be circled back to.

**Comment by davidjgoss** [comment]
@charlierudolph this is ready to go now I think.

In the end it was mostly about changing gherkin document helper to be able to handle the rule/example structure when matching up from pickles, and some usage of this in the JSON formatter - beyond that it's just adding/changing tests on the other formatters to confirm they work normally, and adding some acceptance-level coverage as well.

Let me know if you think anything is missing.

**Comment by davidjgoss** [comment]
@charlierudolph this will probably need a rebase after #1286 goes in, but otherwise is it okay?

File: src/formatter/progress_formatter_spec.ts [hunk 3]
```diff
@@ -31,10 +31,11 @@ Feature: a
...
     Given a passing step
-  Scenario: a5
-    Given a skipped step
-  Scenario: a6
-    Given an undefined step
+  Rule: ab
+    Example: a5
+      Given a skipped step
+    Example: a6
+      Given an undefined step
```
Comment (charlierudolph) [inline src/formatter/progress_formatter_spec.ts]: Can we please make a new test for this vs combining into one? This was really for each step status and I think its useful to have rule as a separate test

File: src/formatter/helpers/gherkin_document_parser_spec.ts [hunk 4]
```diff
@@ -0,0 +1,1294 @@
...
+  getGherkinScenarioMap,
+  getGherkinStepMap,
+} from './gherkin_document_parser'
+
+describe('GherkinDocumentParser', () => {
+  describe('with a Background and Scenario', () => {
+    let sourceData
+    beforeEach(() => {
+      // Arrange
+      sourceData = {
```
Comment (charlierudolph) [inline src/formatter/helpers/gherkin_document_parser_spec.ts]: Really don't like this as we are hardcoding some of gherkin's internal structure in a test that would need to be updated if gherkin changes something. 

Instead of hardcoding it, can we generate this by having the gherkin library parse some content?

Also when asserting the output, can we have maybe only one test assert the full output once to show structure and then later tests can assume the structure is preserved and perform simpler assertions?

Lastly, for general test structure I'm leaning towards no longer using beforeEach and instead just making helper functions so everything you need to know about the test is in the block (instead of having to find beforeEach blocks that setup some state). Thoughts on that?

File: features/fixtures/formatters/failed-rule.json.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+module.exports = [
```
Comment (charlierudolph) [inline features/fixtures/formatters/failed-rule.json.ts]: Do we really need these? I only created json fixtures to make it easy to see if we changed something for backward compatibility. Maybe just one for a rule would be enough?

File: src/formatter/helpers/gherkin_document_parser_spec.ts (hunk 4 above)
Comment (davidjgoss) [reply src/formatter/helpers/gherkin_document_parser_spec.ts]: > Instead of hardcoding it, can we generate this by having the gherkin library parse some content?

Yeah that's much better, will do

> can we have maybe only one test assert the full output once to show structure

Makes sense, will try

> leaning towards no longer using beforeEach and instead just making helper functions

Yeah I see where you're coming from there. I think it would probably improve it in this case. I have seen the `beforeEach` pattern work well where it's use for arrange and act, and then the `it`s that follow it are just a line or two each of asserts.

File: src/formatter/progress_formatter_spec.ts (hunk 3 above)
Comment (davidjgoss) [reply src/formatter/progress_formatter_spec.ts]: Done

File: features/fixtures/formatters/failed-rule.json.ts (hunk 5 above)
Comment (davidjgoss) [reply features/fixtures/formatters/failed-rule.json.ts]: Removed so we are now only adding one json fixture for rule happy path

File: src/formatter/helpers/gherkin_document_parser_spec.ts (hunk 4 above)
Comment (davidjgoss) [reply src/formatter/helpers/gherkin_document_parser_spec.ts]: Okay, I've reworked this now based on the above. So we now parse each example feature with gherkin, and then assert on its output with references - so whilst we are still somewhat coupled to the hierarchical structure of a gherkin document, it will be an obvious compilation error if that changes, rather than assertions being out. Let me know what you think.

**Comment by davidjgoss** [comment]
@charlierudolph feedback addressed and latest changes from master merged in here

File: features/fixtures/formatters/failed-rule.json.ts (hunk 5 above)
Comment (charlierudolph) [reply features/fixtures/formatters/failed-rule.json.ts]: Still seeing failed rule and rejected pickle rule

File: src/formatter/json_formatter.ts [hunk 6]
```diff
@@ -5,15 +5,20 @@ import { formatLocation, GherkinDocumentParser, PickleParser } from './helpers'
...
 import { messages } from 'cucumber-messages'
-import { getGherkinScenarioLocationMap } from './helpers/gherkin_document_parser'
+import {
+  getGherkinExampleRuleMap,
+  getGherkinScenarioLocationMap,
+} from './helpers/gherkin_document_parser'
 import { ITestCaseAttempt } from './helpers/event_data_collector'
 import { doesHaveValue, doesNotHaveValue } from '../value_checker'
 import { parseStepArgument } from '../step_arguments'
+import { io } from 'cucumber-messages/dist/src/cucumber-messages'
```
Comment (charlierudolph) [inline src/formatter/json_formatter.ts]: Please use `import {messages} from 'cucumber-messages'`

File: src/formatter/json_formatter.ts [hunk 7]
```diff
@@ -70,6 +75,9 @@ interface IBuildJsonFeatureOptions {
 interface IBuildJsonScenarioOptions {
   feature: messages.GherkinDocument.IFeature
   gherkinScenarioMap: Dictionary<messages.GherkinDocument.Feature.IScenario>
+  gherkinExampleRuleMap: Dictionary<
+    messages.GherkinDocument.Feature.FeatureChild.IRule
```
Comment (charlierudolph) [inline src/formatter/json_formatter.ts]: Could this be simplified to `IRule`?

File: src/formatter/progress_bar_formatter_spec.ts [hunk 8]
```diff
@@ -94,7 +94,7 @@ describe('ProgressBarFormatter', () => {
         },
         {
           data:
-            'Feature: a\nScenario: b\nGiven a step\nWhen a step\nThen a step',
+            'Feature: a\nRule: b\nExample: c\nGiven a step\nWhen a step\nThen a step',
```
Comment (charlierudolph) [inline src/formatter/progress_bar_formatter_spec.ts]: Can we please create an additional test for this? (and retain the old one)

File: src/formatter/rerun_formatter_spec.ts [hunk 9]
```diff
@@ -180,7 +180,7 @@ describe('RerunFormatter', () => {
                 uri: 'a.feature',
               },
               {
-                data: 'Feature: a\n\nScenario: b\nGiven a step',
+                data: 'Feature: a\n\nRule: b\nExample: c\nGiven a step',
```
Comment (charlierudolph) [inline src/formatter/rerun_formatter_spec.ts]: Same, please add new test

File: src/formatter/helpers/gherkin_document_parser_spec.ts (hunk 4 above)
Comment (charlierudolph) [reply src/formatter/helpers/gherkin_document_parser_spec.ts]: Looks good. Thanks for the update

File: features/fixtures/formatters/failed-rule.json.ts (hunk 5 above)
Comment (davidjgoss) [reply features/fixtures/formatters/failed-rule.json.ts]: Ah okay - I think I thought you were only talking about the fixtures for the JSON formatter. Should I scale back the ones for the message formatter as well?

File: src/formatter/json_formatter.ts (hunk 6 above)
Comment (davidjgoss) [reply src/formatter/json_formatter.ts]: Done

File: src/formatter/json_formatter.ts (hunk 7 above)
Comment (davidjgoss) [reply src/formatter/json_formatter.ts]: Yep, done

File: features/fixtures/formatters/failed-rule.json.ts (hunk 5 above)
Comment (charlierudolph) [reply features/fixtures/formatters/failed-rule.json.ts]: Haha understandable. I meant the fixtures in general (both json + message formatter). Think just a passed version of json / messages is sufficient

File: src/formatter/progress_bar_formatter_spec.ts (hunk 8 above)
Comment (davidjgoss) [reply src/formatter/progress_bar_formatter_spec.ts]: Done

File: src/formatter/rerun_formatter_spec.ts (hunk 9 above)
Comment (davidjgoss) [reply src/formatter/rerun_formatter_spec.ts]: Done

File: features/fixtures/formatters/failed-rule.json.ts (hunk 5 above)
Comment (davidjgoss) [reply features/fixtures/formatters/failed-rule.json.ts]: No worries, done that now.

File: features/formatters.feature [hunk 10]
```diff
@@ -1,17 +1,17 @@
...
+  Scenario: rejected pickle from Scenario
     Given a file named "features/a.feature" with:
       """
       Feature: a feature
         Scenario: a scenario
           Given a step
       """
     When I run cucumber-js with all formatters and `--tags @a`
-    Then the "message" formatter output matches the fixture "formatters/rejected-pickle.message.json"
+    Then the "message" formatter output matches the fixture "formatters/rejected-pickle-scenario.message.json"
```
Comment (charlierudolph) [inline features/formatters.feature]: can we also revert the changes to this scenario and the failed scenario so that we only specify the difference for scenario vs rule for the passed scenarios?

File: src/formatter/progress_formatter_spec.ts [hunk 11]
```diff
@@ -31,7 +31,7 @@ Feature: a
     Given a pending step
   Scenario: a4
     Given a passing step
-  Scenario: a5
+ Scenario: a5
```
Comment (charlierudolph) [inline src/formatter/progress_formatter_spec.ts]: please revert the whitespace change here
